File: backend/app/models/event.py
from datetime import datetime
from typing import Optional, Dict, Any, List
from beanie import Document, Indexed
from pydantic import Field
from pymongo import ASCENDING, DESCENDING, GEOSPHERE


class Event(Document):
    source: Indexed(str)
    domain: Indexed(str)
    event_type: Indexed(str)
    severity: Indexed(float)
    geometry: Dict[str, Any]
    properties: Dict[str, Any] = Field(default_factory=dict)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    timestamp: Indexed(datetime)
    
    class Settings:
        name = "events"
        indexes = [
            [("timestamp", DESCENDING)],
            [("domain", ASCENDING), ("timestamp", DESCENDING)],
            [("severity", DESCENDING), ("timestamp", DESCENDING)],
            [("geometry", GEOSPHERE)],
            [("source", ASCENDING)],
            [("event_type", ASCENDING)],
            [("metadata.severity_tier", ASCENDING)],
        ]

File: frontend/src/types/Event.ts
export interface Event {
  id: string
  source: string
  domain: string
  event_type: string
  severity: number
  geometry: {
    type: 'Point'
    coordinates: [number, number]
  }
  properties: Record<string, any>
  metadata: EventMetadata
  timestamp: string
  created_at: string
}

export interface EventMetadata {
  severity_tier: 'info' | 'low' | 'moderate' | 'high' | 'critical'
  country?: string
  admin1?: string
  city?: string
  timezone?: string
  source_updated?: string
  [key: string]: any
}

export interface EventListResponse {
  events: Event[]
  total: number
  page: number
  page_size: number
}

export interface EventStatsResponse {
  period_hours: number
  domains: Array<{
    domain: string
    count: number
    avg_severity: number
    max_severity: number
  }>
  total: number
}